enterprise/server/remote_execution/dirtools/dirtools.go

```diff
@@ -447,34 +455,71 @@ type FileMap map[digest.Key][]*FilePointer
 
 type BatchFileFetcher struct {
 	ctx          context.Context
+	env          environment.Env
 	instanceName string
-	fileCache    interfaces.FileCache // OPTIONAL
-	bsClient     bspb.ByteStreamClient
-	casClient    repb.ContentAddressableStorageClient // OPTIONAL
+	once         *sync.Once
+	compress     bool
 }
 
 // NewBatchFileFetcher creates a CAS fetcher that can automatically batch small requests and stream large files.
 // `fileCache` is optional. If present, it's used to cache a copy of the data for use by future reads.
 // `casClient` is optional. If not specified, all requests will use the ByteStream API.
-func NewBatchFileFetcher(ctx context.Context, instanceName string, fileCache interfaces.FileCache, bsClient bspb.ByteStreamClient, casClient repb.ContentAddressableStorageClient) *BatchFileFetcher {
+func NewBatchFileFetcher(ctx context.Context, env environment.Env, instanceName string) *BatchFileFetcher {
 	return &BatchFileFetcher{
 		ctx:          ctx,
+		env:          env,
 		instanceName: instanceName,
-		fileCache:    fileCache,
-		bsClient:     bsClient,
-		casClient:    casClient,
+		once:         &sync.Once{},
+		compress:     false,
 	}
 }
 
+func (ff *BatchFileFetcher) supportsCompression() bool {
+	ff.once.Do(func() {
+		if !*enableDownloadCompresssion {
+			return
+		}
+		capabilitiesClient := ff.env.GetCapabilitiesClient()
+		if capabilitiesClient == nil {
+			log.Warningf("Download compression was enabled but no capabilities client found. Cannot verify cache server capabilities")
+			return
+		}
+		enabled, err := cachetools.SupportsCompression(ff.ctx, capabilitiesClient)
+		if err != nil {
+			log.Errorf("Error determinining if cache server supports compression: %s", err)
+		}
+		if enabled {
+			ff.compress = true
+		} else {
+			log.Debugf("Download compression was enabled but remote server did not support compression")
+		}
+	})
+	return ff.compress
+}
+
 func (ff *BatchFileFetcher) batchDownloadFiles(ctx context.Context, req *repb.BatchReadBlobsRequest, filesToFetch FileMap, opts *DownloadTreeOpts) error {
-	if ff.casClient == nil {
+	casClient := ff.env.GetContentAddressableStorageClient()
+	if casClient == nil {
 		return status.FailedPreconditionErrorf("cannot batch download files when casClient is not set")
 	}
 
-	var rsp, err = ff.casClient.BatchReadBlobs(ctx, req)
+	rsp, err := casClient.BatchReadBlobs(ctx, req)
 	if err != nil {
 		return err
 	}
+
+	var decodeBuf []byte
+	if ff.supportsCompression() {
```

Comment:
why don't you pass supportsCompression from the caller? would make the code easier to follow (instead of having two seemingly independent calls to supportsCompression())